src/ehttpc.erl

```diff
@@ -322,76 +367,182 @@ do_request(Client, delete, {Path, Headers}) ->
     gun:delete(Client, Path, Headers).
 
 cancel_stream(Client, StreamRef) ->
-    gun:cancel(Client, StreamRef),
-    flush_stream(Client, StreamRef).
+    %% this is just an async message sent to gun
+    %% the gun stream process does really cancel
+    %% anything, but just mark the receiving process (i.e. self())
+    %% as inactive, however, there could be messages already
+    %% delivered to self()'s mailbox
+    %% or the stream process might send more messages
+    %% before receiving the cancel message.
+    _ = gun:cancel(Client, StreamRef),
+    ok.
 
-flush_stream(Client, StreamRef) ->
-    receive
-        {gun_response, Client, StreamRef, _, _, _} ->
-            flush_stream(Client, StreamRef);
-        {gun_data, Client, StreamRef, _, _} ->
-            flush_stream(Client, StreamRef);
-        {gun_error, Client, StreamRef, _} ->
-            flush_stream(Client, StreamRef)
-	after 0 ->
-		ok
-	end.
-
-await_response(StreamRef, ExpireAt, State = #state{client = Client}) ->
-    case (Timeout = ExpireAt - now_()) > 0 of
-        true ->
-            receive
-                {gun_response, Client, StreamRef, fin, StatusCode, Headers} ->
-                    {reply, {ok, StatusCode, Headers}, State};
-                {gun_response, Client, StreamRef, nofin, StatusCode, Headers} ->
-                    await_remaining_response(StreamRef, ExpireAt, State, {StatusCode, Headers, <<>>});
-                {gun_error, Client, StreamRef, Reason} ->
-                    {reply, {error, Reason}, State};
-                {gun_down, Client, _, Reason, _KilledStreams, _} ->
-                    {reply, {error, Reason}, State};
-                {'DOWN', MRef, process, Client, Reason} ->
-                    true = erlang:demonitor(MRef, [flush]),
-                    NState = case open(State) of
-                                {ok, State1} -> State1;
-                                {error, _Reason} -> State#state{mref = undefined, client = undefined}
-                            end,
-                    {reply, {error, Reason}, NState}
-            after Timeout ->
-                cancel_stream(Client, StreamRef),
-                {reply, {error, timeout}, State}
-            end;
-        false ->
-            cancel_stream(Client, StreamRef),
-            {noreply, State}
+timeout(ExpireAt) ->
+    max(ExpireAt - now_(), 0).
+
+now_() ->
+    erlang:system_time(millisecond).
+
+%% =================================================================================
+%% sent requests
+%% =================================================================================
+
+%% upgrade from old format before 0.1.16
+upgrade_requests(#{pending := _, sent := _} = Already) ->
+    Already;
+upgrade_requests(Map) when is_map(Map) ->
+    #{pending => queue:new(), sent => Map}.
```

Comment:
This should also have `pending_count` and `prioritise_latest`, I think.  Some pattern matches below expect those keys.